src/chat.rs

```diff
@@ -4209,6 +4209,83 @@ pub async fn forward_msgs(context: &Context, msg_ids: &[MsgId], chat_id: ChatId)
     Ok(())
 }
 
+/// Save a copy of the message in "Saved Messages"
+/// and send a sync messages so that other devices save the message as well, unless deleted there.
+pub async fn save_msgs(context: &Context, msg_ids: &[MsgId]) -> Result<()> {
+    for src_msg_id in msg_ids {
+        let dest_rfc724_mid = create_outgoing_rfc724_mid();
+        let src_rfc724_mid = save_copy_in_self_talk(context, src_msg_id, &dest_rfc724_mid).await?;
+        context
+            .add_sync_item(SyncData::SaveMessage {
+                src: src_rfc724_mid,
+                dest: dest_rfc724_mid,
+            })
+            .await?;
+    }
+    context.send_sync_msg().await?;
+    Ok(())
+}
+
+/// Saves a copy of the given message in "Saved Messages" using the given RFC724 id.
+/// To allow UIs to have a "show in context" button,
+/// the copy contains a reference to the original message
+/// as well as to the original chat in case the original message gets deleted.
+/// Returns data needed to add a `SaveMessage` sync item.
+pub(crate) async fn save_copy_in_self_talk(
+    context: &Context,
+    src_msg_id: &MsgId,
+    dest_rfc724_mid: &String,
+) -> Result<String> {
+    let dest_chat_id = ChatId::create_for_contact(context, ContactId::SELF).await?;
+    let mut msg = Message::load_from_db(context, *src_msg_id).await?;
+    msg.param.remove(Param::Cmd);
+    msg.param.remove(Param::WebxdcDocument);
+    msg.param.remove(Param::WebxdcDocumentTimestamp);
+    msg.param.remove(Param::WebxdcSummary);
+    msg.param.remove(Param::WebxdcSummaryTimestamp);
+    msg.param
+        .set_int(Param::OriginalChatId, msg.chat_id.to_u32() as i32);
+    let original_msg_id = if !msg.original_msg_id.is_special() {
+        msg.original_msg_id
+    } else {
+        *src_msg_id
+    };
```

Comment:
What do these lines do, i.e. why isn't `original_msg_id = src_msg_id` in all cases?

It look a bit like they ensure that if you star a message that was already starred in the Saved Messages chat, the original message will be starred again? I think that this would be confusing, it would be better if the UI didn't offer any button to star the message again and instead offered to un-star it also when long-tapping it in Saved Messages (that's what WA does, too).